### PR TITLE
test: add core tests for lib and contracts

### PR DESCRIPTION
**Pull request by slongdotexe**
# Purpose
Improve test coverage across both web2 and web3 libs.

# Testing
- Web2 lib can be tested with `yarn test` in `packages/sde-bitmaps-lib`
- Web3 lib can be tested with `yarn hardhat test` in `packages/contracts`

### CODE REVIEW

File: packages/contracts/contracts/implementations/SdeBitmapsCore.sol [hunk 1]
```diff
@@ -53,7 +72,7 @@ library SdeBitmapsCore {
    */
   function setBit(BitMap storage bitmap, uint256 bitIndex) internal {
     uint256 bucket = bitIndex >> 8;
```
Comment (mjLabrys) [inline packages/contracts/contracts/implementations/SdeBitmapsCore.sol:74]: Sir I need you to use getBucketFromBitIndex

File: packages/contracts/contracts/implementations/SdeBitmapsCore.sol [hunk 2]
```diff
@@ -63,10 +82,45 @@ library SdeBitmapsCore {
...
+  ) internal {
+    uint256 bucket = bitmap._data[bucketIndex];
+    bitmap._data[bucketIndex] = bucket >> places;
+  }
+
+  // ##### UNTESTED CODE #####
+
   // TODO: tests
   /**
    * @dev Sets the bit indices in `indeices` in the bucket at `bucketIndex`.
```
Comment (mjLabrys) [inline packages/contracts/contracts/implementations/SdeBitmapsCore.sol:126]: indeices

**Review by mjkid221 — COMMENTED** [review]
You know what would be big brain, is if you convert all of these to assembly code once the contract's finished haha. Something like this

`function getBucketFromBitIndex(uint256 bitIndex) internal pure returns (uint256 bucket) {
    assembly {
        bucket := shr(8, bitIndex)
    }
}`
I guess a simple operation like this won't benefit much but could play around with the more complex ones

**Comment by sjustintaylor** [comment]
> You know what would be big brain, is if you convert all of these to assembly code once the contract's finished haha. Something like this
> 
> `function getBucketFromBitIndex(uint256 bitIndex) internal pure returns (uint256 bucket) { assembly { bucket := shr(8, bitIndex) } }` I guess a simple operation like this won't benefit much but could play around with the more complex ones

Some real big brain stuff right here

**Review by sjustintaylor — APPROVED** [review]

**Comment by slongdotexe** [comment]
Yeah, I had considered in the long term that writing direct assembly would likely lead to performance optimisations for more complex stuff (e.g. PRBMath). 
As @mjLabrys pointed out, the current operations are relatively basic, so the compiler will likely fully optimise them. However, if more complex operations are included in the future, this might be a handy gas improvement.

File: packages/contracts/contracts/implementations/SdeBitmapsCore.sol (hunk 2 above)
Comment (slongdotexe) [reply packages/contracts/contracts/implementations/SdeBitmapsCore.sol:126]: 7nzkkr

File: packages/contracts/contracts/implementations/SdeBitmapsCore.sol (hunk 1 above)
Comment (slongdotexe) [reply packages/contracts/contracts/implementations/SdeBitmapsCore.sol:74]: I took this opportunity to condense some of the boiler-plate which should also reduce stack memory usage too.

**Pull request merged**